File: ts/web3js/src/instructions/eata/initializeGlobalVault.ts
import {
  PublicKey,
  TransactionInstruction,
  SystemProgram,
  AccountMeta,
} from "@solana/web3.js";
import { EATA_PROGRAM_ID } from "../../constants";
import { globalVaultPdaWithBumpFromMint } from "../../pda";

export interface InitializeGlobalVaultAccounts {
  payer: PublicKey;
  mint: PublicKey;
}

export function createInitializeGlobalVaultInstruction(
  accounts: InitializeGlobalVaultAccounts,
): TransactionInstruction {
  const [vault, bump] = globalVaultPdaWithBumpFromMint(accounts.mint);

  const keys: AccountMeta[] = [
    { pubkey: vault, isWritable: true, isSigner: false },
    { pubkey: accounts.payer, isWritable: true, isSigner: true },
    { pubkey: accounts.mint, isWritable: false, isSigner: false },
    { pubkey: SystemProgram.programId, isWritable: false, isSigner: false },
  ];

  const buffer = Buffer.alloc(2);
  buffer[0] = 1; // discriminator
  buffer[1] = bump;

  return new TransactionInstruction({
    programId: EATA_PROGRAM_ID,
    keys,
    data: buffer,
  });
}

File: rust/sdk/src/spl/instructions/initialize_global_vault.rs
use crate::{
    consts::ESPL_TOKEN_PROGRAM_ID,
    solana_compat::solana::{system_program, AccountMeta, Instruction, Pubkey},
    spl::EphemeralSplDiscriminator,
};

/// Create an initialize global vault instruction.
pub fn initialize_global_vault(
    payer: Pubkey,
    vault: Pubkey,
    mint: Pubkey,
    vault_bump: u8,
) -> Instruction {
    Instruction {
        program_id: ESPL_TOKEN_PROGRAM_ID,
        accounts: vec![
            AccountMeta::new(vault, false),
            AccountMeta::new(payer, false),
            AccountMeta::new_readonly(mint, false),
            AccountMeta::new_readonly(system_program::id(), false),
        ],
        data: vec![
            EphemeralSplDiscriminator::InitializeGlobalVault as u8,
            vault_bump,
        ],
    }
}
